src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -407,5 +406,176 @@ public static void SetLastWriteTime(string fullPath, DateTimeOffset time, bool a
+                    out _,
+                    IntPtr.Zero);
+
+                if (!success)
+                {
+                    int error = Marshal.GetLastWin32Error();
+                    // The file or directory is not a reparse point.
+                    if (error == Interop.Errors.ERROR_NOT_A_REPARSE_POINT)
+                    {
+                        return null;
+                    }
+
+                    throw Win32Marshal.GetExceptionForWin32Error(error, linkPath);
+                }
+
+                ReadOnlySpan<byte> bufferSpan = new(buffer);
+                ref readonly Interop.Kernel32.REPARSE_DATA_BUFFER rdb = ref MemoryMarshal.AsRef<Interop.Kernel32.REPARSE_DATA_BUFFER>(bufferSpan);
+
+                // Only symbolic links are supported at the moment.
+                if ((rdb.ReparseTag & Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK) == 0)
+                {
+                    return null;
+                }
+
+                int substituteNameOffset = sizeof(Interop.Kernel32.REPARSE_DATA_BUFFER) + rdb.ReparseBufferSymbolicLink.SubstituteNameOffset;
+                int substituteNameLength = rdb.ReparseBufferSymbolicLink.SubstituteNameLength;
+
+                ReadOnlySpan<char> targetPath = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(substituteNameOffset, substituteNameLength));
+
+                // Target path is relative, we need to append the link directory.
+                if (normalize && (rdb.ReparseBufferSymbolicLink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) != 0)
+                {
+                    return Path.Join(Path.GetDirectoryName(linkPath.AsSpan()), targetPath);
+                }
+
+                return targetPath.ToString();
+            }
+            finally
+            {
+                if (buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(buffer);
+                }
+            }
+        }
+
+        private static unsafe string? GetFinalLinkTarget(string linkPath, bool isDirectory)
+        {
+            Interop.Kernel32.WIN32_FIND_DATA data = default;
+            GetFindData(linkPath, isDirectory, ref data);
+
+            // The file or directory is not a reparse point.
+            if ((data.dwFileAttributes & (uint)FileAttributes.ReparsePoint) == 0 ||
+                // Only symbolic links are supported at the moment.
+                (data.dwReserved0 & Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK) == 0)
+            {
+                return null;
+            }
+
+            using SafeFileHandle handle = OpenSafeFileHandle(linkPath,
+                    Interop.Kernel32.FileOperations.OPEN_EXISTING |
+                    Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS);
+
+            if (handle.IsInvalid)
+            {
+                Win32Marshal.GetExceptionForLastWin32Error(linkPath);
+            }
+
+            char* buffer = stackalloc char[Interop.Kernel32.MAX_PATH];
```

Comment:
Can it be long path?

Reply:
I discussed this with @Jokzee. We should verify what happens if we pass a path that's larger than `MAX_PATH`. if the API can fail and tell us that the buffer was not large enough, then we could apply similar logic to what we added in the Unix side.

Reply:
@iSazonov : It can.